src/algebra/group/basic.lean

```diff
@@ -1,76 +1,328 @@
 /-
 Copyright (c) 2014 Jeremy Avigad. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Jeremy Avigad, Simon Hudon, Mario Carneiro
+Authors: Jeremy Avigad, Leonardo de Moura, Simon Hudon, Mario Carneiro
 -/
 import algebra.group.to_additive
+import tactic.simpa
 import logic.function.basic
 
-attribute [simp] sub_neg_eq_add
+set_option default_priority 100
+set_option old_structure_cmd true
+
+universe u
+
+/- Additive "sister" structures.
+   Example, add_semigroup mirrors semigroup.
+   These structures exist just to help automation.
+   In an alternative design, we could have the binary operation as an
+   extra argument for semigroup, monoid, group, etc. However, the lemmas
+   would be hard to index since they would not contain any constant.
+   For example, mul_assoc would be
+
+   lemma mul_assoc {α : Type u} {op : α → α → α} [semigroup α op] :
+                   ∀ a b c : α, op (op a b) c = op a (op b c) :=
+    semigroup.mul_assoc
+
+   The simplifier cannot effectively use this lemma since the pattern for
+   the left-hand-side would be
+
+        ?op (?op ?a ?b) ?c
 
-/-!
-# Extra identities for semigroups, monoids, and groups
+   Remark: we use a tactic for transporting theorems from the multiplicative fragment
+   to the additive one.
 -/
 
-universes u v w
-variables {M : Type u} {A : Type v} {G : Type w}
+@[ancestor has_mul] class semigroup (α : Type u) extends has_mul α :=
+(mul_assoc : ∀ a b c : α, a * b * c = a * (b * c))
+@[ancestor has_add] class add_semigroup (α : Type u) extends has_add α :=
+(add_assoc : ∀ a b c : α, a + b + c = a + (b + c))
+attribute [to_additive add_semigroup] semigroup
+attribute [to_additive add_semigroup.to_has_add] semigroup.to_has_mul
+attribute [to_additive add_semigroup.add_assoc] semigroup.mul_assoc
+
+section semigroup
+variables {α : Type u} [semigroup α]
+
+@[no_rsimp, to_additive]
+lemma mul_assoc : ∀ a b c : α, a * b * c = a * (b * c) :=
+semigroup.mul_assoc
+attribute [no_rsimp] add_assoc -- TODO(Mario): find out why this isn't copying
```

Comment:
Because `no_rsimp` is not in the list of attributes copied by `to_additive`.

Reply:
Answer: `no_rsimp` is not a basic attribute, so `tactic.copy_attribute` doesn't work. What should I use instead?

Reply:
I don't have any suggestions, but if you are looking into this, `ematch` also needs to be copied.